core/src/main/java/org/apache/iceberg/hadoop/HadoopCatalog.java

```diff
@@ -200,7 +211,108 @@ public void renameTable(TableIdentifier from, TableIdentifier to) {
   }
 
   @Override
-  public void close() throws IOException {
+  public void createNamespace(Namespace namespace, ImmutableMap<String, String> meta) {
+    Preconditions.checkArgument(
+        !namespace.isEmpty(),
+        "Cannot create namespace with invalid name: %s", namespace);
+    Preconditions.checkArgument(
+        meta.isEmpty(),
+        "Hadoop Catalog not support metadata %s on namespace: %s", meta, namespace);
+
+    Path nsPath = new Path(SLASH.join(warehouseLocation, SLASH.join(namespace.levels())));
+    FileSystem fs = Util.getFs(nsPath, conf);
+
+    try {
+      if (isNamespace(fs, nsPath)) {
+        throw new AlreadyExistsException("Namespace '%s' already exists!", namespace);
+      }
+
+      fs.mkdirs(nsPath);
+
+    } catch (IOException e) {
+      throw new RuntimeIOException(e, "Create namespace failed: %s", namespace);
+    }
+  }
+
+  @Override
+  public List<Namespace> listNamespaces(Namespace namespace) {
+    Path nsPath = new Path(SLASH.join(warehouseLocation, SLASH.join(namespace.levels())));
+    FileSystem fs = Util.getFs(nsPath, conf);
+
+    if (!isNamespace(fs, nsPath)) {
+      throw new NoSuchNamespaceException("Namespace does not exist: %s", namespace);
+    }
+
+    try {
+      if (!namespace.isEmpty()) {
```

Comment:
Instead of using two paths here, I think this should use a helper method that appends the new name to the namespace that is listed, like this:

```java
static Namespace append(Namespace ns, String name) {
  String[] levels = Arrays.copy(ns.levels(), 0, ns.levels().length + 1);
  levels[ns.levels().length] = name;
  return Namespace.of(levels);
}
```